solidity_bridge/src/Relayer.sol

```diff
@@ -1,41 +1,68 @@
 pragma solidity ^0.8.24;
 
 import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
-import {IProover} from "./interfaces/IProover.sol";
+import {IProver} from "./interfaces/IProver.sol";
 import {IRelayer} from "./interfaces/IRelayer.sol";
 import {Constants} from "./libraries/Constants.sol";
 
 
 
 contract Relayer is IRelayer, AccessControl {
-    IProover private _proover;
+    IProver private _prover;
     mapping(uint256=>bytes32) private _block_numbers;
     mapping(bytes32=>uint256) private _merkle_roots;
 
+    uint256 private constant P = 2**64 - 2**32 + 1;
+    uint256 private constant MASK_52BITS = (2**52) - 1;
 
 
-    function initialize(address proover) external {
+    function initialize(address prover) external {
         if(getRoleAdmin(Constants.ADMIN_ROLE) != DEFAULT_ADMIN_ROLE) revert AlreadyInitialized();
         _setRoleAdmin(Constants.ADMIN_ROLE, Constants.ADMIN_ROLE);
         _grantRole(Constants.ADMIN_ROLE, msg.sender );
-        _proover = IProover(proover);
+        _prover = IProver(prover);
     }
 
-    function add_merkle_root(uint256 blockNumber, bytes32 merkle_root, bytes calldata proof ) external {
-        bytes memory message = abi.encodePacked(blockNumber, merkle_root);
-        if(!_proover.verifyProof(message, proof)) {
+    function add_merkle_root_with_block(uint256 block_number, bytes32 merkle_root, bytes calldata proof ) external {
```

Comment:
You need public inputs anyway as, for example, `block_number` may be `100` and may be `100 + p`. This difference matters only for verifier as it'll not verify `block_number = 100` and `block_number = 100 + p` at the same time, but for `Relayer` contract `100` and `100 + p` mean the same

Reply:
I suggest to leave only `add_merkle_root_with_inputs` fn